chainerx_cc/chainerx/routines/n_step_rnn.cc

```diff
@@ -0,0 +1,500 @@
+    std::vector<Array> bs_5_8{bs[6], bs[4], bs[5], bs[7]};
+    Array hb = StackWeight(bs_5_8);
+
+    Array lstm_in = Linear(x, xw, xb) + Linear(h, hw, hb);
+
+    std::vector<Array> lstm_out = Lstm(*c, lstm_in);
+
+    return lstm_out;
+}
+
+std::vector<Array> RnnImpl(
+        const Array& x,
+        const Array& h,
+        const absl::optional<Array>& c,
+        const std::vector<Array>& ws,
+        const std::vector<Array>& bs,
+        const absl::optional<std::string>& activation) {
+    c.has_value();
+    Array xw = ws[0];
+    Array hw = ws[1];
+
+    Array xb = bs[0];
+    Array hb = bs[1];
+
+    Array rnn_in_1 = Linear(x, xw, xb);
+    Array rnn_in_2 = Linear(h, hw, hb);
+    Array rnn_in = rnn_in_1 + rnn_in_2;
+    std::vector<Array> out{};
+    out.reserve(1);
+    Array rnn_act;
+    if (*activation == "tanh") {
+        rnn_act = Tanh(rnn_in);
+    } else {
+        rnn_act = Relu(rnn_in);
+    }
+    out.emplace_back(rnn_act);
+    return out;
+}
+
+template <typename Impl>
+std::vector<std::vector<Array>> OneDirectionalLoop(
+        Impl&& impl,
+        std::vector<Array>& xs,
+        Array h,
+        absl::optional<Array> c,
+        const std::vector<Array>& ws,
+        const std::vector<Array>& b,
+        absl::optional<std::string> activation) {
+    Shape h_shape{h.shape()[1], h.shape()[2]};
+    h = Reshape(h, h_shape);
+    if (c.has_value()) {
+        *c = Reshape(*c, h_shape);
+    }
+    std::vector<Array> h_list;
+    for (auto& x : xs) {
+        if (x.shape()[0] > h.shape()[0]) {
+            throw DimensionError{"The batch size of x must be equal to or less than the size of state", x.shape(), ' ', h.shape()};
+        }
+        std::vector<int64_t> indices_h;
+        indices_h.emplace_back(x.shape()[0]);
+        indices_h.emplace_back(h.shape()[0]);
+        std::vector<Array> h_split = Split(h, indices_h, 0);
+        std::vector<Array> c_split;
+        std::vector<Array> h_c;
+        if (c.has_value()) {
+            std::vector<int64_t> indices_c;
+            indices_c.emplace_back(x.shape()[0]);
+            indices_c.emplace_back(c->shape()[0]);
+            c_split = Split(*c, indices_c, 0);
+            h_c = impl(x, h_split[0], c_split[0], ws, b, activation);
+        } else {
+            h_c = impl(x, h_split[0], c, ws, b, activation);
+        }
+
+        h_list.emplace_back(h_c[0]);
+        h_split[0] = h_c[0];
+        if (c.has_value()) {
+            c_split[0] = h_c[1];
+            c = Concatenate(c_split, 0);
+        }
+        h = Concatenate(h_split, 0);
+    }
+    std::vector<std::vector<Array>> out;
+    std::vector<Array> state;
+    state.emplace_back(h);
+    if (c.has_value()) {
+        state.emplace_back(*c);
+    }
+    out.emplace_back(state);
+    out.emplace_back(h_list);
+
+    return out;
+}
+
+template <typename Impl>
+std::vector<std::vector<Array>> NStepRnnImpl(
+        Impl&& impl,
+        int64_t n_layers,
+        Array hx,
```

Comment:
There are clant-tidy errors.

I think this can be `const Array&` and `std::move` can be removed.

Reply:
I have made the changes. Could you check?